src/Compilers/CSharp/Portable/Symbols/ConstraintsHelper.cs

```diff
@@ -352,21 +352,29 @@ internal static class ConstraintsHelper
             this Symbol containingSymbol,
             Binder binder,
             ImmutableArray<TypeParameterSymbol> typeParameters,
+            TypeParameterListSyntax typeParameterList,
             SyntaxList<TypeParameterConstraintClauseSyntax> constraintClauses,
             Location location,
             DiagnosticBag diagnostics)
         {
-            if (typeParameters.Length == 0 || constraintClauses.Count == 0)
+            if (typeParameters.Length == 0)
             {
                 return ImmutableArray<TypeParameterConstraintClause>.Empty;
             }
 
+            if (constraintClauses.Count == 0)
+            {
+                ImmutableArray<TypeParameterConstraintClause> defaultClauses = binder.GetDefaultTypeParameterConstraintClauses(typeParameterList);
+
+                return defaultClauses.ContainsOnlyEmptyConstraintClauses() ? ImmutableArray<TypeParameterConstraintClause>.Empty : defaultClauses;
```

Comment:
>defaultClauses.ContainsOnlyEmptyConstraintClauses() [](start = 23, length = 51)

Should `Binder.GetDefaultTypeParameterConstraintClauses()` simply return `ImmutableArray<TypeParameterConstraintClause>.Empty` if all are empty?

Reply:
> Should Binder.GetDefaultTypeParameterConstraintClauses() simply return ImmutableArray<TypeParameterConstraintClause>.Empty if all are empty?

Some consumers do not want this optimization.

---
In reply to: [273625867](https://github.com/dotnet/roslyn/pull/34797#discussion_r273625867) [](ancestors = 273625867)